### PR TITLE
Add a way for Resolver consumers to cancel operational resolve attempts.

### LINKED ISSUE #23881: Darwin leaks DNSServiceResolve calls
If a resolve in the sense of DNSServiceResolve does not complete, we never cancel it.  We can keep accumulating them until things fall over.

Proposed plan for fixing this:

1. Remove the IPAddressType argument from Dnssd::Resolver::ResolveNodeId.
2. Add a CancelResolve (name TBD) API on Dnssd::Resolver, which takes a PeerId.  This will be called from the default AddressResolve::Resolver impl when things are canceled or time out.
3. Clearly document the expectation that Dnssd::Resolver implementations must count the number of interested consumers for an operational lookup attempt, treating ResolveNodeId as a signal to increase the consumer count and treating CancelResolve as a signal to decrease the consumer count.
4. ResolveNodeId should be documented to kick an existing lookup to try more aggressively if it was doing backoff.  Minimal mdns seems to be doing this already.
5. ChipDnssdResolve should be documented as allowing coalescing of resolves as long as the context, and peer id are the same and as long as each call kicks things that are doing backoff as needed.
6. Add ChipDnssdCancelResolve that takes a context and peer id.

Then we can clean things up when consumers no longer care about them.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
Adds a way for consumers to notify Resolver when they no longer care about an operational resolve, so a Resolver implementation can keep track of how many consumers are interested and stop work as desired if no one is interested.

Fixes https://github.com/project-chip/connectedhomeip/issues/23881

### CODE REVIEW

File: src/lib/dnssd/Discovery_ImplPlatform.cpp [hunk 1]
```diff
@@ -625,6 +625,15 @@ CHIP_ERROR DiscoveryImplPlatform::ResolveNodeId(const PeerId & peerId)
...
 
+void DiscoveryImplPlatform::NodeIdResolutionNoLongerNeeded(const PeerId & peerId)
+{
+    char name[Common::kInstanceNameMaxLength + 1];
+    if (MakeInstanceName(name, sizeof(name), peerId) == CHIP_NO_ERROR)
+    {
+        ChipDnssdResolveNoLongerNeeded(name);
+    }
+}
+
```
Comment (vivien-apple) [inline src/lib/dnssd/Discovery_ImplPlatform.cpp:636]: nit: 
```
void DiscoveryImplPlatform::NodeIdResolutionNoLongerNeeded(const PeerId & peerId)
{
    char name[Common::kInstanceNameMaxLength + 1];
    ReturnOnFailure(MakeInstanceName(name, sizeof(name), peerId));

    ChipDnssdResolveNoLongerNeeded(name);
}
```

File: src/lib/dnssd/Discovery_ImplPlatform.cpp [hunk 2]
```diff
@@ -671,15 +680,30 @@ CHIP_ERROR ResolverProxy::ResolveNodeId(const PeerId & peerId)
...
+    return err;
+}
+
+void ResolverProxy::NodeIdResolutionNoLongerNeeded(const PeerId & peerId)
+{
+    char name[Common::kInstanceNameMaxLength + 1];
+    if (MakeInstanceName(name, sizeof(name), peerId) == CHIP_NO_ERROR)
+    {
+        ChipDnssdResolveNoLongerNeeded(name);
+    }
```
Comment (vivien-apple) [inline src/lib/dnssd/Discovery_ImplPlatform.cpp:706]: nit:
```
void ResolverProxy::NodeIdResolutionNoLongerNeeded(const PeerId & peerId)
{
    char name[Common::kInstanceNameMaxLength + 1];
    ReturnOnFailure(MakeInstanceName(name, sizeof(name), peerId));

    ChipDnssdResolveNoLongerNeeded(name);
}

```

File: src/lib/dnssd/Discovery_ImplPlatform.cpp [hunk 3]
```diff
@@ -671,15 +680,30 @@ CHIP_ERROR ResolverProxy::ResolveNodeId(const PeerId & peerId)
...
+
+    mDelegate->Retain();
+
+    CHIP_ERROR err = ChipDnssdResolve(&service, Inet::InterfaceId::Null(), HandleNodeIdResolve, mDelegate);
+    if (err != CHIP_NO_ERROR)
+    {
+        mDelegate->Release();
+    }
+    return err;
+}
```
Comment (vivien-apple) [inline src/lib/dnssd/Discovery_ImplPlatform.cpp:698]: All of this change is unrelated right ? Looks like a bug fix of us beeing to zealous to retain the delegate.

File: src/platform/Darwin/DnssdImpl.cpp [hunk 4]
```diff
@@ -344,7 +344,23 @@ static CHIP_ERROR Resolve(void * context, DnssdResolveCallback callback, uint32_
     ChipLogDetail(Discovery, "Resolve type=%s name=%s interface=%" PRIu32, StringOrNullMarker(type), StringOrNullMarker(name),
                   interfaceId);
 
-    auto sdCtx = chip::Platform::New<ResolveContext>(context, callback, addressType);
+    // This is a little silly, in that resolves for the sane name, type, etc get
```
Comment (vivien-apple) [inline src/platform/Darwin/DnssdImpl.cpp:347]:
```suggestion
    // This is a little silly, in that resolves for the same name, type, etc get
```

File: src/platform/Darwin/DnssdImpl.cpp [hunk 5]
```diff
@@ -344,7 +344,23 @@ static CHIP_ERROR Resolve(void * context, DnssdResolveCallback callback, uint32_
...
+    // coalesced by the underlying mDNSResponder anyway.  But we need to keep
+    // track of our context/callback/etc, (even though in practice it's always
+    // exactly the same) and the interface id (which might actually be different
+    // for different Resolve calls).  So for now just keep using a
```
Comment (vivien-apple) [inline src/platform/Darwin/DnssdImpl.cpp:351]:
```suggestion
    // for different Resolve calls). So for now just keep using a
```

File: src/platform/Darwin/DnssdImpl.cpp [hunk 6]
```diff
@@ -450,6 +471,34 @@ CHIP_ERROR ChipDnssdResolve(DnssdService * service, chip::Inet::InterfaceId inte
...
+    auto existingCtx = MdnsContexts::GetInstance().GetExistingResolveForInstanceName(instanceName);
+    if (!existingCtx)
+    {
+        return;
+    }
+
+    if (*existingCtx->consumerCounter == 0)
+    {
+        return;
+    }
```
Comment (vivien-apple) [inline src/platform/Darwin/DnssdImpl.cpp:485]: nit:
```
VerifyOrReturn(nullptr != existingCtx);
VerifyOrReturn(*exisingCtx->consumerCounter != 0);
```

File: src/platform/Darwin/DnssdImpl.cpp [hunk 7]
```diff
@@ -450,6 +471,34 @@ CHIP_ERROR ChipDnssdResolve(DnssdService * service, chip::Inet::InterfaceId inte
...
 
+void ChipDnssdResolveNoLongerNeeded(const char * instanceName)
+{
+    auto existingCtx = MdnsContexts::GetInstance().GetExistingResolveForInstanceName(instanceName);
+    if (!existingCtx)
+    {
+        return;
+    }
+
+    if (*existingCtx->consumerCounter == 0)
```
Comment (vivien-apple) [inline src/platform/Darwin/DnssdImpl.cpp:482]: This should not happens right ?

**Review by vivien-apple — APPROVED** [review]

File: src/lib/dnssd/Discovery_ImplPlatform.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/Discovery_ImplPlatform.cpp:636]: Good idea, done.

File: src/lib/dnssd/Discovery_ImplPlatform.cpp (hunk 2 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/Discovery_ImplPlatform.cpp:706]: Done.

File: src/platform/Darwin/DnssdImpl.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/platform/Darwin/DnssdImpl.cpp:347]: Nice catch!  Fixed.

File: src/platform/Darwin/DnssdImpl.cpp (hunk 6 above)
Comment (bzbarsky-apple) [reply src/platform/Darwin/DnssdImpl.cpp:485]: Done.

File: src/lib/dnssd/ActiveResolveAttempts.cpp [hunk 8]
```diff
@@ -92,6 +92,22 @@ CHIP_ERROR ActiveResolveAttempts::CompleteAllBrowses()
...
 }
 
+void ActiveResolveAttempts::NodeIdResolutionNoLongerNeeded(const PeerId & peerId)
+{
+    for (auto & item : mRetryQueue)
+    {
+        if (item.attempt.Matches(peerId))
+        {
+            item.attempt.ConsumerRemoved();
+            if (item.attempt.ResolveData().consumerCount == 0)
```
Comment (andy31415) [inline src/lib/dnssd/ActiveResolveAttempts.cpp:102]: should this piece of code be inside ConsumeRemoved? since all of the calls (check in ResolveData and Clear) are done inside `item.attempt` it seems that we can put everything inside `ActiveResolveAttempt`

File: src/lib/dnssd/ActiveResolveAttempts.h [hunk 9]
```diff
@@ -181,6 +186,28 @@ class ActiveResolveAttempts
         bool IsIpResolve() const { return resolveData.Is<IpResolve>(); }
         void Clear() { resolveData = DataType(); }
 
+        void WillCoalesceWith(const ScheduledAttempt & existing)
+        {
+            if (IsResolve() && existing.Matches(*this))
```
Comment (andy31415) [inline src/lib/dnssd/ActiveResolveAttempts.h:191]: This method needs some documentation about what its purpose and use is.
I am also wondering of the `Matches` is not met - what does that mean and should we silently ignore the overwrite? What sets consumerCount to 0/1 to start with?

Also to decrease indent and help documentation we could split ifs:

```cpp
if (!IsResolve()) {
    return; // consumer count is only tracked for resolve requests
}

if (!existing.Matches(*this)) {
   // Adding another consumer to the same query, 
   // TODO: what if this does NOT match? is this overwriting? should we log an error?
   return; 
}

resolveData.Get<Resolve>().consumerCount++;
```

File: src/lib/dnssd/ActiveResolveAttempts.h [hunk 10]
```diff
@@ -181,6 +186,28 @@ class ActiveResolveAttempts
...
+            if (!IsResolve())
+            {
+                return;
+            }
+
+            auto & count = resolveData.Get<Resolve>().consumerCount;
+            if (count > 0)
+            {
+                --count;
+            }
```
Comment (andy31415) [inline src/lib/dnssd/ActiveResolveAttempts.h:208]: would the else branch warrant a log (is it an internal error to remove a consumer when count is already thought to be 0?)

File: src/platform/Darwin/DnssdImpl.cpp (hunk 7 above)
Comment (bzbarsky-apple) [reply src/platform/Darwin/DnssdImpl.cpp]: Ah, this is about the other check, the one before we decrement the counter.

This _could_ happen with some weird and unexpected access patterns: if someone starts a new resolve for an instance name while being notified about a completed resolve for that instance name.  The guard is there to ensure we don't leak, at least, in that situation.

File: src/lib/dnssd/ActiveResolveAttempts.cpp (hunk 8 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/ActiveResolveAttempts.cpp:102]: Ah, indeed.  That is a good idea.

File: src/lib/dnssd/ActiveResolveAttempts.h (hunk 9 above)
Comment (bzbarsky-apple) [reply src/lib/dnssd/ActiveResolveAttempts.h:191]: > This method needs some documentation about what its purpose and use is.

Will do.

> I am also wondering of the Matches is not met - what does that mean and should we silently ignore the overwrite? 

If it's not met, the new attempt is for a different resolve and it has a different set of consumers.  We already log an error for this case in `ActiveResolveAttempts::MarkPending`:
```
        // TODO: node was evicted here, if/when resolution failures are
        // supported this could be a place for error callbacks
        //
        // Note however that this is NOT an actual 'timeout' it is showing
        // a burst of lookups for which we cannot maintain state. A reply may
        // still be received for this peer id (query was already sent on the
        // network)
        ChipLogError(Discovery, "Re-using pending resolve entry before reply was received.");
```

I will add some better documentation.

> What sets consumerCount to 0/1 to start with?

The `Resolve` struct constructor, since it's declared as:
```
            uint32_t consumerCount = 0;
```

> Also to decrease indent and help documentation we could split ifs:

Done.

File: src/lib/dnssd/ActiveResolveAttempts.h [hunk 11]
```diff
@@ -68,7 +69,11 @@ class ActiveResolveAttempts
             IpResolve(HeapQName && host) : hostName(std::move(host)) {}
         };
 
-        ScheduledAttempt() {}
+        ScheduledAttempt()
+        {
+            static_assert(sizeof(Resolve) <= sizeof(Browse) || sizeof(Resolve) <= sizeof(HeapQName),
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/lib/dnssd/ActiveResolveAttempts.h:74]: It's checking that I did not increase the size of ScheduledAttempt by adding that member.  I'll make the message clearer.

File: src/lib/dnssd/Resolver.h [hunk 12]
```diff
@@ -369,11 +369,31 @@ class Resolver
...
+     * @see NodeIdResolutionNoLongerNeeded.
      */
     virtual CHIP_ERROR ResolveNodeId(const PeerId & peerId) = 0;
 
+    /*
+     * Notify the resolver that one of the consumers that called ResolveNodeId
+     * successfully no longer needs the resolution result (e.g. because it got
+     * the result, or got an error, or no longer cares about future updates).
+     * There must be a NodeIdResolutionNoLongerNeeded call that matches every
+     * successful ResolveNodeId call.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/lib/dnssd/Resolver.h:393]: Added some clarifying comments.

File: src/lib/dnssd/ActiveResolveAttempts.h (hunk 10 above)
Comment (andy31415) [reply src/lib/dnssd/ActiveResolveAttempts.h:226]: I would comment here and maybe log if we try to decrease a 0 count (unless it really happens easily). I expect a pairing where we say 'add ref' and 'dec ref' to not really generally go negative.

**Review by andy31415 — APPROVED** [review]
